src/preprocess.rkt

```diff
@@ -1,44 +1,122 @@
+  (define query (make-egg-query (cons expression (append evens* swaps*)) rules))
+  (match-define (cons expression~ rest~) (map last (simplify-batch query)))
+  (define-values (evens~ swaps~) (split-at rest~ (length evens*)))
+  (define swaps (filter-map
+                 (lambda (pair swap~) (and (equal? expression~ swap~) pair))
+                 pairs
+                 swaps~))
+  (define components (connected-components variables swaps))
+  (define abs-instructions
+    (for/list ([variable (in-list variables)] [even~ (in-list evens~)]
+               #:when (equal? expression~ even~))
+      (list 'abs variable)))
+  (define sort-instructions
+    (for/list ([component (in-list components)]
+               #:when (> (length component) 1))
+      (cons 'sort component)))
+  ;; Absolute value should happen before sorting
+  (append abs-instructions sort-instructions))
+
+(define (connected-components variables swaps)
+  (define components (disjoint-set (length variables)))
+  (for ([swap (in-list swaps)])
+    (match-define (list a b) swap)
+    (disjoint-set-union! components
+                         (disjoint-set-find! components (index-of variables a))
+                         (disjoint-set-find! components (index-of variables b))))
+  (group-by
+   (compose (curry disjoint-set-find! components) (curry index-of variables))
+   variables))
+
+(define (preprocess-pcontext context pcontext preprocessing)
+  (define preprocess
+    (apply compose (map
+                    (curry instruction->operator context)
+                    ;; Function composition applies the rightmost function first
+                    (reverse preprocessing))))
+  (pcontext-map preprocess pcontext))
+
+(define (instruction->operator context instruction)
+  (define variables (context-vars context))
+  (define sort* (curryr sort (curryr </total (context-repr context))))
+  (match instruction
+    [(list 'sort component ...) #:when (equal? component variables)
+     sort*]
+    [(list 'sort component ...) #:when (list-prefix? component variables)
+     (define position (length component))
+     (lambda (points)
+       (let-values ([(prefix suffix) (split-at points position)])
+         (append (sort* prefix) suffix)))]
+    [(list 'sort component ...) #:when (list-suffix? component variables)
+     (define position (- (length variables) (length component)))
+     (lambda (points)
+       (let-values ([(prefix suffix) (split-at points position)])
+         (append prefix (sort* suffix))))]
+    [(list 'sort component ...)
+     (unless (subsequence? component variables)
+       (error 'instruction->operator
+              "component should always be a subsequence of variables"))
+     (define indices (indexes-where variables (curryr member component)))
+     (lambda (points)
+       (let* ([subsequence (list-ref* points indices)]
+              [sorted (sort* subsequence)])
+         (list-set* points indices sorted)))]
```

Comment:
Ok, c'mon, what's this all about? Why not just use the `list-set*` case always?

Reply:
Haha, I noticed that many of our sorting steps were either the entire list of variables or a prefix of it, so I added these as fast cases where the list only gets looped through once for sort instead of three times. Happy to get rid of them though. I can also see if a nested `cond` would look better.

Reply:
Although, now I'm realizing that with the way lists work I think it is actually still going through the list several times because `split-at` and `append` are both linear. I will remove it.

Reply:
Ok, so "fast cases" are only going to worth it if it's speeding up something that:
1. Runs a lot of times, or
2. Is typically slow
What's your guess, does this qualify? No—it has at most like 15 inputs, so even O(n^2) is going to be fast enough. And it runs once per benchmark.